File: 0074-Search-a-2D-Matrix/lc0074.rs
// Rust solution.

use std::cmp::Ordering;

impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        if !matrix.is_empty(){
            let (m, n) = (matrix.len(), matrix[0].len());
            let (mut lo, mut hi) = (0_usize, m*n);
            while lo < hi {
                let mid = (lo + hi) >> 1;
                match matrix[mid/n][mid%n].cmp(&target) {
                    Ordering::Less    => { lo = mid+1; },
                    Ordering::Greater => { hi = mid; },
                    Ordering::Equal   => { return true; },
                }
            }
        }
        false
    }
}

File: 0074-Search-a-2D-Matrix/lc0074.cc
// C++ solution.

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if (matrix.empty()) {
            return false;
        }
        int m = matrix.size(), n = matrix[0].size();
        int lo = 0, hi = m*n;
        while (lo < hi) {
            int mid = lo + ((hi - lo) >> 1);
            int x = matrix[mid/n][mid%n];
            if (x < target) {
                lo = mid + 1;
            } else if (x > target) {
                hi = mid;
            } else {
                return true;
            }
        }
        return false;
    }
};
